### PR TITLE
Wasm integration

### PR DESCRIPTION
**Pull request by rogueloop**
## What type of MR is this? (check all applicable)
- [ x] 🍕 Feature

### Web assembly integration for Lama2. 
Modifying Golang to add support for building web assembly binary.

## Important files to start review from in the order

1. [controller.cli.go](https://github.com/HexmosTech/Lama2/pull/47/files#diff-e695ec0a8b7f510f1ffd1f13c8d49a01ee6fef63b1a6f8c337d7b5c6e654febb)

2. [Controller.wasm.go](https://github.com/HexmosTech/Lama2/pull/47/files#diff-5c8c095d751d4a6ca04027cdc4129b26b19f8d4394748b300e2d60d39fbe4ee0)

3. [controller.go](https://github.com/HexmosTech/Lama2/pull/47/files#diff-e695ec0a8b7f510f1ffd1f13c8d49a01ee6fef63b1a6f8c337d7b5c6e654febb)

### other major changes :
* makefile

```shell
# Build command for normal cli builds
go build -tags=cli -o build/l2 -ldflags "-X main.version=$(git tag --sort=-version:refname | head -n 1) -X main.lamawasm=cli " l2.go

# Build command for wasm builds
GOOS=js GOARCH=wasm go build -o build/l2.wasm -tags=wasm -ldflags "-X main.version=$(git tag --sort=-version:refname | head -n 1) -X main.lamawasm=lamawasm " l2.wasm.go
```

### CODE REVIEW

File: .github/workflows/testapp.yml [hunk 1]
```diff
@@ -16,11 +16,11 @@ jobs:
       - name: Install dependencies
         run: go mod tidy
       - name: Build
-        run: go build -v ./...
+        run: go build -tags=cli -v ./...
```
Comment (RijulTP) [inline .github/workflows/testapp.yml:19]: Is this done only for ones with tag cli, are there any other tags

File: cmdgen/cmdgen.go [hunk 2]
```diff
@@ -70,7 +55,8 @@ func assembleCmdString(httpv string, url string, jsonObj *gabs.Container, header
 
 	if multipart {
 		for key, val := range jsonObj.Data().(*gabs.Container).ChildrenMap() {
-			command = append(command, "'"+key+"'='"+val.Data().(string)+"'  ")
+			keyValuePair := fmt.Sprintf("%s=%s", key, val.Data().(string))
+			command = append(command, keyValuePair)
```
Comment (RijulTP) [inline cmdgen/cmdgen.go:59]: Are these changes tested?

File: .github/workflows/testapp.yml (hunk 1 above)
Comment (rogueloop) [reply .github/workflows/testapp.yml:19]: yes there is `wasm` tag also currently build is failing , checking that now

File: controller/controller.wasm.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	script := b.Data().(string)
+	js.Global().Call("eval", script)
+}
+
+func ExecuteRequestorBlock(block *gabs.Container) httpie.ExResponse {
+	preprocess.ProcessVarsInBlock(block)
+	// TODO - replace stuff in headers, and varjson and json as well
+	cmd, stdinBody := cmdgen.ConstructCommand(block)
+	resp, e1 := cmdexec.ExecCommand(cmd, stdinBody)
+	printFields(resp)
```
Comment (RijulTP) [inline controller/controller.wasm.go:38]: What does this do

File: l2.wasm.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	if len(version) == 0 {
+		version = "vUnset"
+	}
+}
+
+func main() {
+	// Set up the WebAssembly function
+	js.Global().Set("lama2Wasm", js.FuncOf(lama2Wasm))
+	js.Global().Set("goWebRequestFunc", wasmLamaPromise())
+	select {}
```
Comment (RijulTP) [inline l2.wasm.go:26]: Can add a comment on what these lines do

File: l2.wasm.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+func lama2Wasm(this js.Value, i []js.Value) interface{} {
+	go func() {
+		controller.ProcessWasmInput("GET\nhttps://httpbin.org/get")
```
Comment (RijulTP) [inline l2.wasm.go:31]: Is this for processing the Default request?

File: l2lsp/methods/lsp_lifecycle.go [hunk 6]
```diff
@@ -6,7 +6,7 @@ import (
 
 	"github.com/HexmosTech/lama2/l2lsp/request"
 	"github.com/HexmosTech/lama2/l2lsp/response"
-	"github.com/rs/zerolog/log"
+	// "github.com/rs/zerolog/log"
```
Comment (RijulTP) [inline l2lsp/methods/lsp_lifecycle.go:9]: This can be removed right

File: l2.wasm.go (hunk 5 above)
Comment (rogueloop) [reply l2.wasm.go:31]: That function was for testing purpose , removed

File: preprocess/preprocess.wasm.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+func ProcessVarsInBlock(block *gabs.Container) {
+	ExpandURL(block)
+	ExpandHeaders(block)
+	ExpandJSON(block)
+}
+
+func ExpandHeaders(block *gabs.Container) {
+	headerMap := block.S("details", "headers")
+	// log.Info().Str("HeaderMap", headerMap.String()).Msg("")
```
Comment (RijulTP) [inline preprocess/preprocess.wasm.go:32]: Can remove commented code

File: preprocess/preprocess.wasm.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
+//go:build wasm
```
Comment (RijulTP) [inline preprocess/preprocess.wasm.go:1]: What is the basic difference in the code between preprocess.wasm.go and preprocess.go

File: l2lsp/methods/lsp_lifecycle.go (hunk 6 above)
Comment (rogueloop) [reply l2lsp/methods/lsp_lifecycle.go:9]: this is for logging purpose

File: controller/controller.wasm.go (hunk 3 above)
Comment (rogueloop) [reply controller/controller.wasm.go:38]: Added it for debugging purpose, printing the response

File: preprocess/preprocess.wasm.go (hunk 8 above)
Comment (rogueloop) [reply preprocess/preprocess.wasm.go:1]: there is a lot of code repetition there converted it into 3 separate file to remove duplication

Comment (rogueloop) [reply preprocess/preprocess.wasm.go:1]: The difference is in function calls , wasm has 1 parameter , cli has 2 params , each of them calls different functions

File: docs/Lama2/site/tutorials/main.wasm
Comment (lovestaco) [inline docs/Lama2/site/tutorials/main.wasm:1]: Why is the wasm file aswell pushed to the repo?

Comment (lovestaco) [reply docs/Lama2/site/tutorials/main.wasm:1]: Cant the doc site use the injector or something?

File: docs/Lama2/site/main.wasm
Comment (lovestaco) [inline docs/Lama2/site/main.wasm:1]: One more bin here both 24mb

File: cmdexec/cmdexec.wasm.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+// ExecCommand changes directory to the given `apiDir`
+// and then executes the command specified in `cmdStr`
+// During command execution, ExecCommand streams output
+// to stdout.
+// Once execution finishes, previous CWD is restored,
+// and the command output is returned as a string
+func ExecCommand(cmdSlice []string, stdinBody string) (httpie.ExResponse, error) {
+	proxyURL := "https://proxyserver.hexmos.com/"	
+	proxyUserName := "proxyServer"
+	proxyUserPassword := "proxy22523146server"
```
Comment (lovestaco) [inline cmdexec/cmdexec.wasm.go:27]: Passwords are open, variables are not declared for the URL, name, password. Is the password meant to be open?

**Comment by lovestaco** [comment]
Too many files need an abstract PR description, like 
1. A wasm bin is generated
2. The bin starts from funcname() from filename.ext
3. Moves to blah for doing X
4. Moves to blah2 for doing Y

It would be better to write in this way, so that when you're writing documentation you can reuse this and iterate on that.

File: cmdexec/js.wasm.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"syscall/js"
+)
+
+func RunVMCode(jsCode string) {
+	js.Global().Call("eval", jsCode)
+}
+
+func GenerateChainCode(httpRespBody string) string {
```
Comment (LinceMathew) [inline cmdexec/js.wasm.go:13]: can you add a comment on what this function is doing? what is  generate chain code?

File: cmdgen/cmdgen.wasm.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/HexmosTech/gabs/v2"
+)
+
+func assembleCmdString(httpv string, url string, jsonObj *gabs.Container, headers *gabs.Container, multipart bool, form bool) ([]string, string) {
```
Comment (LinceMathew) [inline cmdgen/cmdgen.wasm.go:14]: add a comment about what is this function doing

File: utils/utils.wasm.go [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+		return err
+	}
+
+	return nil
+}
+
+// UpdateSelf downloads the installation script from
+// the official repository, and executes it to update
+// the l2 binary to the latest version
+func UpdateSelf() {
```
Comment (LinceMathew) [inline utils/utils.wasm.go:159]: is this function for a self-update wasm file or a L2 binary?

File: l2lsp/methods/lsp_suggestEnv.go [hunk 13]
```diff
@@ -9,7 +9,6 @@ import (
 	l2envpackege "github.com/HexmosTech/lama2/l2env"
 	"github.com/HexmosTech/lama2/l2lsp/request"
 	"github.com/HexmosTech/lama2/l2lsp/response"
-	"github.com/rs/zerolog/log"
 )
```
Comment (shrsv) [inline l2lsp/methods/lsp_suggestEnv.go:13]: Have we tested the VSCode extension with new CLI build? Worth doing it once I think.

**Pull request merged**